datafusion/physical-expr/src/conditional_expressions.rs

```diff
@@ -35,36 +33,51 @@ pub fn coalesce(args: &[ColumnarValue]) -> Result<ColumnarValue> {
         )));
     }
 
-    let size = match args[0] {
-        ColumnarValue::Array(ref a) => a.len(),
-        ColumnarValue::Scalar(ref _s) => 1,
-    };
-    let mut res = new_null_array(&args[0].data_type(), size);
+    let return_type = args[0].data_type();
+    let mut return_array = args.iter().filter_map(|x| match x {
+        ColumnarValue::Array(array) => Some(array.len()),
+        _ => None,
```

Comment:
why not use the same logic by returning `Some(1)`?

Reply:
return_array is used to decide whether to return a scalar or a vector, so if we see no array sizes and get None from `return_array.next()` we know all params are scalar and will enter branch in line-70 and do fast scalar value coalesce.